Use the right Swift ObjectUtil class depending on parent class

mrackwitz opened the pull request:
`RealmSwiftObjectUtil` doesn't support `RLMObject` classes (and shouldn't from my understanding). That can cause that just linking Realm Swift screws up `RLMObject` model definitions from Swift as seen in #3097.
This introduces the protocol `ObjectProtocol` implemented by RealmSwiftObject and SwiftFakeObject  and tests for the conformance to that as a workaround to allow invalid test classes.
/c @jpsim @tgoyne

tgoyne commented:
This makes it so that Swift-defined RLMObject subclasses are not treated as Swift classes in some of the places that they need to be.

mrackwitz commented:
@tgoyne: The only places which are concerned from this changeset are within `RLMObjectSchema` as the property `isSwiftClass` continues to reflect where the class was defined instead what superclass it is inheriting from. Can you please point out where exactly you think something goes wrong?

tgoyne commented:
Swift-defined RLMObject subclasses need to use initSwiftPropertyWithName for RLMArray properties, and the Any/String? error message applies to them.

mrackwitz commented:
Ah right, so I would pass both `isSwiftClass` and `isSwiftObject` into the method to change the handling there? Alternatively, we could have a further function on base of `RLMObjectUtil` which takes a class and goes over the runtime, but that could mean a slight performance degradation in comparison.

tgoyne commented:
Is there any obvious reason it wouldn't work to have a class method on RLMObject and RealmSwift.Object (and the fake base classes) which returns the appropriate ObjectUtil class? It seems like that'd both simpler and faster than what the code is currently doing. The only downside I can think of is that it might not be possible to fully prevent the Swift version from showing up in autocomplete (the obj-c version just needs to not be in the public header).

mrackwitz commented:
That works fine, but it pesters indeed Realm Swift's public API.

mrackwitz commented:
I addressed what @tgoyne proposed here and rebased it now. So I'm waiting now for CI feedback and would merge it then after adding a changelog item, if no one objects.

bdash commented:
Can a test case be written for this?

mrackwitz commented:
Our current test harness doesn't really allow testing failures at link time.
Theoretically we could let our tests for Realm Objective-C link also again Realm Swift, but that could allow that other issues could be introduced.
